Clean up mode handling in python dispatcher

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121444
* __->__ #121083

Things that were bad before this PR:
1. Temporarily unsetting functional tensor mode and proxy mode both had duplicate implementation
2. There are variants of mode handling private utils that has duplicate implementation. (different APIs calling repeated implementation, so i refactored)
3. _push_mode API used to take dispatch key argument which is not necessary.
4. There are unused APIs.

tugsbayasgalan committed:
Clean up mode handling in python dispatcher

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Clean up mode handling in python dispatcher"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Clean up mode handling in python dispatcher"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Clean up mode handling in python dispatcher"




[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/map.py:
```diff
@@ -13,8 +13,8 @@
 from torch._ops import HigherOrderOperator
 from torch._subclasses.fake_tensor import FakeTensorMode
 from torch._subclasses.functional_tensor import (
+    disable_functional_mode,
```
Do you have some explanations for what was cleaned up and why? It's a bit hard to tell what improved

tugsbayasgalan replied:
Updated

tugsbayasgalan committed:
Update on "Clean up mode handling in python dispatcher"


Things that were bad before this PR:
1. Temporarily unsetting functional tensor mode and proxy mode both had duplicate implementation
2. There are variants of mode handling private utils that has duplicate implementation. (different APIs calling repeated implementation, so i refactored)
3. _push_mode API used to take dispatch key argument which is not necessary.
4. There are unused APIs. 



[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge
